src/core/common/appender.hpp

```diff
@@ -0,0 +1,186 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *  This file defines OpenThread `Appender` class.
+ */
+
+#ifndef APPENDER_HPP_
+#define APPENDER_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/data.hpp"
+#include "common/message.hpp"
+#include "common/type_traits.hpp"
+
+namespace ot {
+
+/**
+ * The `Appender` class acts as a wrapper over either a `Message` or a data buffer (e.g. `MutableData`) and provides
+ * different flavors of `Append()` method.
+ *
+ * This class helps in construction of message content where the destination can be either a `Message` or a buffer.
+ *
+ */
+class Appender
+{
+public:
+    /**
+     * This enumeration represent the `Appender` Type (whether appending to a `Message` or data buffer).
+     *
+     */
+    enum Type : uint8_t
+    {
+        kMessageAppender, ///< `Appender` appends to a `Message`
+        kBufferAppender,  ///< `Appender` appends to a buffer (`MutableData`).
```

Comment:
nit: including `Appender` in type name is redundant?
```suggestion
        kMessage, ///< `Appender` appends to a `Message`
        kBuffer,  ///< `Appender` appends to a buffer (`MutableData`).
```